src/SAD/Parser/Token.hs

```diff
@@ -72,60 +79,69 @@ isProperToken EOF{} = True
-    posToken pos whitespaceBefore mode s | not (Text.null lexem) = tok:toks
+    -- Activate the tokenizer when '\begin{forthel}' appears.
+    posToken :: TexState -> SourcePos -> TokenType -> Text -> [Token]
+    posToken OutsideForthelEnv pos _ s = toks
+      where
+        (ignoredText, rest) = Text.breakOn "\\begin{forthel}" s
+        newPos = advanceAlong pos (ignoredText <> "\\begin{forthel}")
+        toks = posToken InsideForthelEnv newPos WhiteSpaceBefore (Text.drop 15 rest)
+    
+    -- Deactivate the tokenizer when '\end{forthel}' appears.
+    posToken InsideForthelEnv pos _ s | start == "\\end{forthel}" = toks
+      where
+        (start,rest) = Text.splitAt 13 s
+        toks = posToken OutsideForthelEnv (advanceAlong pos start) WhiteSpaceBefore rest
+        
+    -- Make alphanumeric tokens that don't start with whitespace.
+    posToken texState pos whitespaceBefore s | not (Text.null lexeme) = tok:toks
       where
-        (lexem, rest) = Text.span isLexem s
-        tok  = makeToken lexem pos whitespaceBefore mode
-        toks = posToken (advanceAlong pos lexem) NoWhiteSpaceBefore mode rest
-    posToken pos _ mode s | not (Text.null white) = toks
+        (lexeme, rest) = Text.span isLexeme s
+        tok  = makeToken lexeme pos whitespaceBefore
+        toks = posToken texState (advanceAlong pos lexeme) NoWhiteSpaceBefore rest
+    
+    -- Process whitespace.
+    posToken texState pos _ s | not (Text.null white) = toks
       where
         (white, rest) = Text.span isSpace s
-        toks = posToken (advanceAlong pos white) WhiteSpaceBefore mode rest
-    posToken pos whitespaceBefore mode s = case Text.uncons s of
+        toks = posToken texState (advanceAlong pos white) WhiteSpaceBefore rest
+    
+    -- Process non-alphanumeric symbol or EOF.
+    posToken texState pos whitespaceBefore s = case Text.uncons s of
       Nothing -> [EOF pos]
       Just ('\\', rest) -> tok:toks
         where
           (name, rest') = Text.span isAlpha rest
-          cmd = (Text.cons '\\' name)
-          tok = makeToken cmd pos whitespaceBefore mode
-          toks = posToken (advanceAlong pos cmd) WhiteSpaceBefore mode rest'
-      Just ('$', rest) -> toks
-        where
-          toks = posToken (advancePos pos '$') NoWhiteSpaceBefore (switchMode mode) rest
-          switchMode :: TokenMode -> TokenMode
-          switchMode TextMode = MathMode
-          switchMode MathMode = TextMode
-      Just ('#', _) -> tok:toks
-        where
-          (comment, rest) = Text.break (== '\n') s
-          tok  = makeToken comment pos Comment mode
-          toks = posToken (advanceAlong pos comment) whitespaceBefore mode rest
-      Just ('%', _) -> tok:toks
+          cmd = Text.cons '\\' name
+          tok = makeToken cmd pos whitespaceBefore
+          toks = posToken texState (advanceAlong pos cmd) WhiteSpaceBefore rest'
+      Just (c, _) | if usingTexParser texState then c == '%' else c `elem` ['%','#'] -> tok:toks
```

Comment:
Ah, so '%' is still valid in non-tex texts for comment blocks?

Reply:
Apparently I forgot to remove `%`. This has been fixed now.